nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/excel/RowIterator.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */package org.apache.nifi.excel;
+
+import com.github.pjfanning.xlsx.StreamingReader;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.ss.usermodel.Workbook;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class RowIterator implements Iterator<Row>, Closeable {
+    private final Workbook workbook;
+    private final Iterator<Sheet> sheets;
+    private Sheet currentSheet;
+    private Iterator<Row> currentRows;
+    private final Map<String, Boolean> desiredSheets;
+    private final int firstRow;
+    private ComponentLog logger;
+    private boolean log;
+    private Row currentRow;
+
+    public RowIterator(InputStream in, List<String> desiredSheets, int firstRow) {
+        this(in, desiredSheets, firstRow, null);
+    }
+
+    public RowIterator(InputStream in, List<String> desiredSheets, int firstRow, ComponentLog logger) {
+        this.workbook = StreamingReader.builder()
+                .rowCacheSize(100)
+                .bufferSize(4096)
+                .open(in);
+        this.sheets = this.workbook.iterator();
+        this.desiredSheets = desiredSheets != null ? desiredSheets.stream()
+                .collect(Collectors.toMap(key -> key, value -> Boolean.FALSE)) : new HashMap<>();
+        this.firstRow = firstRow;
+        this.logger = logger;
+        this.log = logger != null;
+    }
+
+    @Override
+    public boolean hasNext() {
+        setCurrent();
+        boolean next = currentRow != null;
+        if(!next) {
+            String sheetsNotFound = getSheetsNotFound(desiredSheets);
+            if (!sheetsNotFound.isEmpty() && log) {
+                logger.warn("Excel sheet(s) not found: {}", sheetsNotFound);
+            }
```

Comment:
This check and warning seems unnecessary in general, perhaps there are other places this could be logged if needed.

Reply:
This the bulletin generated which is remnant from the ConvertExcelToCsv processor. I thought this is the only place to log it as it is the only data structure that iterates the sheets of the Excel document.

Reply:
Some of the other Record Readers add a record.count attribute, which is useful for tracking. That's probably a better option than logging a warning.

Reply:
I thought it may be useful for operators to know explicitly when a sheet did not exist. I personally think that is clearer than a record count which could be misleading.

Reply:
@exceptionfactory I did not see other record readers with a record.count attribute. Which one(s) did  you see this in? I did a search for `record,.count`

Reply:
@exceptionfactory Does the exception have to be specifically an `IOException`? The problem I have is if that or any other checked exception is thrown then the `hasNext`  method signature has to change and that clashes with the` java.util.Iterator` interface. Would a `RuntimeException` be okay to throw?

Reply:
It could be something else, probably `ProcessException` would be a good option as a more specific subclass of `RuntimeException`.

Reply:
Perfect, thanks!